tokio-util/src/net/mod.rs

```diff
@@ -0,0 +1,104 @@
+//! TCP/UDP/Unix helpers for tokio.
+
+use crate::either::Either;
+use std::future::Future;
+use std::io::Result;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+#[cfg(unix)]
+pub mod unix;
+
+/// A trait for a listener: `TcpListener` and `UnixListener`.
+pub trait Listener {
+    /// The stream's type of this listener.
+    type Io: tokio::io::AsyncRead + tokio::io::AsyncWrite;
+    /// The socket address type of this listener.
+    type Addr;
+
+    /// Polls to accept a new incoming connection to this listener.
+    fn poll_accept(&mut self, cx: &mut Context<'_>) -> Poll<Result<(Self::Io, Self::Addr)>>;
+
+    /// Accepts a new incoming connection from this listener.
+    fn accept(&mut self) -> ListenerAcceptFut<'_, Self>
+    where
+        Self: Sized,
+    {
+        ListenerAcceptFut::new(self)
+    }
+
+    /// Returns the local address that this listener is bound to.
+    fn local_addr(&self) -> Result<Self::Addr>;
+}
+
+impl Listener for tokio::net::TcpListener {
+    type Io = tokio::net::TcpStream;
+    type Addr = std::net::SocketAddr;
+
+    fn poll_accept(&mut self, cx: &mut Context<'_>) -> Poll<Result<(Self::Io, Self::Addr)>> {
+        Self::poll_accept(self, cx)
+    }
+
+    fn local_addr(&self) -> Result<Self::Addr> {
+        self.local_addr().map(Into::into)
+    }
+}
+
+/// Future for accepting a new connection from a listener.
+#[derive(Debug)]
+pub struct ListenerAcceptFut<'a, L> {
+    listener: &'a mut L,
+}
+
+impl<'a, L> ListenerAcceptFut<'a, L>
+where
+    L: Listener,
+{
+    fn new(listener: &'a mut L) -> Self {
+        Self { listener }
+    }
+}
+
+impl<'a, L> Future for ListenerAcceptFut<'a, L>
+where
+    L: Listener,
+{
+    type Output = Result<(L::Io, L::Addr)>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.listener.poll_accept(cx)
+    }
+}
+
+impl<L, R> Listener for Either<L, R>
+where
+    L: Listener,
+    R: Listener,
+{
+    type Io = Either<<L as Listener>::Io, <R as Listener>::Io>;
+    type Addr = Either<<L as Listener>::Addr, <R as Listener>::Addr>;
+
+    fn poll_accept(&mut self, cx: &mut Context<'_>) -> Poll<Result<(Self::Io, Self::Addr)>> {
+        match self {
+            Either::Left(listener) => listener
+                .poll_accept(cx)
+                .map(|res| res.map(|(stream, addr)| (Either::Left(stream), Either::Left(addr)))),
+            Either::Right(listener) => listener
+                .poll_accept(cx)
+                .map(|res| res.map(|(stream, addr)| (Either::Right(stream), Either::Right(addr)))),
+        }
+    }
```

Comment:
Hmm. It's unfortunate that the types here allow it return stuff like `(Left(io), Right(addr))` even though it wont in practice. Ideas?

Reply:
Wow I'm so blind I didn't even notice.

Yes you're right in practice this will never happen. If someones use `match` on the Result, they will need to make unreachable branches just for that. It's stupid.

.... :thinking: I will think about it... but I don't see a solution right now.

Reply:
`EitherListener` would not implement `Listener`, it would have its own method with their own signatures

Reply:
You could just put the method on the `Either` we already have.

Reply:
Another option is to change the trait so that `Addr` and `Io` are merged into a single type which each listener will define as a pair.

Reply:
> Another option is to change the trait so that Addr and Io are merged into a single type which each listener will define as a pair.

That would work too but not super elegant I guess...

> You could just put the method on the Either we already have.

Good point

Reply:
Ok I replaced the Listener impl on Either by custom methods here: b55b62c3

Unfortunately the accept Future struct is slightly different so I had to duplicate some code.

Reply:
Why not just make it an async fn and drop the poll function? We don't need these tricks when it's not a trait method.

Reply:
:exploding_head:

Reply:
Done in 316f015e